rasa/core/actions/action.py

```diff
@@ -158,30 +148,40 @@ def is_retrieval_action(action_name: Text, retrieval_intents: List[Text]) -> boo
 
 
 def action_from_name(
-    name: Text,
-    action_endpoint: Optional[EndpointConfig],
-    user_actions: List[Text],
-    should_use_form_action: bool = False,
-    retrieval_intents: Optional[List[Text]] = None,
+    name: Text, domain: Domain, action_endpoint: Optional[EndpointConfig]
 ) -> "Action":
-    """Return an action instance for the name."""
+    """Retrieves an action by its name.
 
+    Args:
+        name: The name of the action.
+        domain: The current model domain.
+        action_endpoint: The endpoint to execute custom actions.
+
+    Returns:
+        The instantiated action.
+    """
     defaults = {a.name(): a for a in default_actions(action_endpoint)}
 
-    if name in defaults and name not in user_actions:
+    if name in defaults and name not in domain.user_actions_and_forms:
         return defaults[name]
-    elif name.startswith(UTTER_PREFIX) and is_retrieval_action(
-        name, retrieval_intents or []
+
+    if name.startswith(UTTER_PREFIX) and is_retrieval_action(
+        name, domain.retrieval_intents
     ):
         return ActionRetrieveResponse(name)
-    elif name.startswith(UTTER_PREFIX):
+
+    if name.startswith(UTTER_PREFIX):
         return ActionUtterTemplate(name)
-    elif should_use_form_action:
+
+    is_form = name in domain.form_names
+    # Users can override the form by defining an action with the same name as the form
+    user_overrode_form_action = is_form and name in domain.user_actions
+    if is_form and not user_overrode_form_action:
```

Comment:
```python
    if is_form and not user_overrode_form_action:
```
is equivalent to (`X and not (X and Y)`):
```python
    if is_form and not (is_form and name in domain.user_actions):
```
which in turn is equivalent to (`X and not Y`):
```python
    if is_form and name not in domain.user_actions:
```
So then maybe we could just write:
```python
    if name in domain.form_names and name not in domain.user_actions:
```

Reply:
I assigned names to the single checks to make their meaning more explicit.

Developers might wonder what `name not in domain.user_actions` means while `user_overrode_form_action` conveys way more information about this. What do you think?